# Bioinformatics_Final_Paper
Final project report for Bioinformatics (BIOL 5340)
I used ChatGPT to help explain what three error messages I received while making code in VSC.
First prompt: What do these messages mean? File "/home/administrator/mambaforge/bin/xpclr", line 196, in <module> main() File "/home/administrator/mambaforge/bin/xpclr", line 87, in main assert os.access(outdir, os.W_OK), \ AssertionError: No permission to write in the specified directory:
ChatGPT's response: These messages are a Python traceback coming from the xpclr program, and they‚Äôre telling you exactly why the program crashed.
What the Error Means:
1. xpclr calls main()
File "/home/administrator/mambaforge/bin/xpclr", line 196, in <module>
    main()


This just says the script is running its main function.

2. The error is triggered inside main()
File "/home/administrator/mambaforge/bin/xpclr", line 87, in main
    assert os.access(outdir, os.W_OK), \


xpclr is checking whether it has permission to write to the output directory you provided with --out or --outdir.

‚ùå 3. The assertion failed
AssertionError: No permission to write in the specified directory:


This means:

xpclr does NOT have permission to write to the output folder you gave.
Code that I used: Option 1: Create the directory first
mkdir -p /path/to/outdir

Option 2: Give yourself permission
chmod u+w /path/to/outdir


Or for all users:

chmod 777 /path/to/outdir


(only if you understand the security implications)

I verified these three options by running them in bash, once I tried all three of these options, my xpclr analysis was able to work.

Second prompt: What does this message mean? Could not load .tbi/.csi index of Scut.Vir.Hybrid.combined.cohort.118.samples.20231002.WITH.CNV.QC.fullfiltered.snps.g.phased.run1.vcf.gz: No such file or directory); falling back to scanning to region
ChatGPT output: This message is coming from a tool trying to access the index file for a compressed VCF (.vcf.gz). Here's what it means in plain English:

‚úÖ Meaning of the Error

‚ÄúCould not load .tbi/.csi index ‚Ä¶ No such file or directory; falling back to scanning to region‚Äù

This means:

Your VCF file is bgzip-compressed (.vcf.gz), and the software expects an index file:

either yourfile.vcf.gz.tbi

or yourfile.vcf.gz.csi

The program looked for that index but could not find it in the same directory.

Because the index is missing, the tool cannot quickly jump to a region inside the VCF, so it says:
‚Äúfalling back to scanning‚Äù:
‚Üí meaning it will now read through the entire VCF file line-by-line instead of efficiently seeking.

This will still work (most of the time), but it is much slower.
üõ†Ô∏è How to Fix It

You need to create an index file for your VCF.

Option 1 ‚Äî Using tabix (most common)
tabix -p vcf Scut.Vir.Hybrid.combined.cohort.118.samples.20231002.WITH.CNV.QC.fullfiltered.snps.g.phased.run1.vcf.gz


This will create:

Scut.Vir.Hybrid.combined.cohort.118.samples.20231002.WITH.CNV.QC.fullfiltered.snps.g.phased.run1.vcf.gz.tbi

I attempted to verify this output solution by running the code on my intended code, creating the ".tbi" extension. This did not solve the problem of running the xpclr package. A simpler solution was given to me by my labmate.

Third prompt: What do you put for the value if you want all of the chromosomes? 
- the xpclr package does not allow you to iterate over all of the chromosomes in your file, you must iterate over one at a time. I need to perform the xpclr test on all of my chromosomes.

ChatGPT output: XP-CLR does NOT accept ‚Äúall chromosomes‚Äù as a value for --chr.
You must run XP-CLR separately for each chromosome/contig.

There is no valid value like:

--chr all
--chr *
--chr 1-12


XP-CLR requires exactly one contig name per run.
‚úÖ How to run XP-CLR on all chromosomes
1. Get a list of chromosome names from your VCF

Run:

bcftools view -h your.vcf.gz | grep "^##contig" | cut -f1 -d"," | sed 's/##contig=<ID=//'


This prints something like:

chr1
chr2
chr3
...

2. Loop over chromosomes and run XP-CLR for each
Bash loop example
for CHR in chr1 chr2 chr3 chr4 chr5 chr6 chr7 chr8 chr9 chr10 chr11 chr12; do
    xpclr \
        --input xpclr_input.txt \
        --map genetic_map.txt \
        --pop1 parental.scut.txt \
        --pop2 parental.vir.txt \
        --chr $CHR \
        --out results.$CHR.txt
done


Or a fully automated version:

for CHR in $(bcftools view -h your.vcf.gz | grep "^##contig" | cut -f1 -d"," | sed 's/##contig=<ID=//'); do
    xpclr --input xpclr_input.txt \
          --map genetic_map.txt \
          --pop1 parental.scut.txt \
          --pop2 parental.vir.txt \
          --chr $CHR \
          --out xpclr.$CHR.txt
done
I was unable to verify this ouput, as my labmate pointed out that there is an easier solution where you did not have to list every chromosome in your file. I ended up using a bash script that stored all of the information into a variable, which saved a lot time.
